[c10d] Add a recursive method to get the inner most store

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117093
* #117092
* #116545
* #117075
* __->__ #117074

In c10d PG initialization, we wrap TCPStore with multiple layers of PrefixStore which adds layers of prefix.

One example is: 
"default_pg/0//cuda//timeout_dump"
When initialized the default PG, because there is no store passed. We first add the prefix "default_pg" to the TCPStore returned from rendezvous:

https://github.com/pytorch/pytorch/blob/bdeaaad70ccc5a34ef6244b6454d01ca2c582645/torch/distributed/distributed_c10d.py#L1240

We then add pg_name (aka 0) https://github.com/pytorch/pytorch/blob/bdeaaad70ccc5a34ef6244b6454d01ca2c582645/torch/distributed/distributed_c10d.py#L1376 and device (aka cuda) https://github.com/pytorch/pytorch/blob/bdeaaad70ccc5a34ef6244b6454d01ca2c582645/torch/distributed/distributed_c10d.py#L1387

to the prefix. Then when we call store_->set("timeout_dump"). The actual key used for writing into TCPStore is "default_pg/0//cuda//timeout_dump".

For sub-PG, things get even interesting, we put the store wrapped with default pg name to a cache:
https://github.com/pytorch/pytorch/blob/bdeaaad70ccc5a34ef6244b6454d01ca2c582645/torch/distributed/distributed_c10d.py#L1517

And when creating each subPG, it is append its PG name right after the cached store. The example keys are:
'default_pg/0//10//cuda//timeout_dump', 'default_pg/0//12//cuda//timeout_dump', 'default_pg/0//38//cuda//timeout_dump', 'default_pg/0//39//cuda//timeout_dump'. (10, 12, 38 and 39 are all PG names of each subPG created)

The reason why the number in the name is bumped up so high is because for each subPG creation, all ranks have to call the API together and the global variable used for PG name will be bumped up monolithically:

https://github.com/pytorch/pytorch/blob/bdeaaad70ccc5a34ef6244b6454d01ca2c582645/torch/distributed/distributed_c10d.py#L3666

Similar things happen for using hashing for PG names.

This has a potential issue, because each sub-PG has an instance of ProcessGroupNCCL, and if we want to set something global to notify all sub-PGs (and all ranks). This added prefix causes bugs. For example, if on sub-PG 1, we set a value to TCPStore with key ('default_pg/0//1//cuda//timeout_dump'), while we use the default PG instances to check the TCPStore, which are using the key ('default_pg/0//cuda//timeout_dump'), default PG instances will never get the notified signals. So in this PR, we added a new API in PrefixStore which we get the innermost non-PrefixStore for set and check. The next PR will make changes in NCCL watchdog.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @wz337 @tianyu-l @wconstab @yf225

fduwjj committed:
[c10d] Add a recursive method to get the inner most store

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/PrefixStore.cpp:
```diff
@@ -108,4 +108,19 @@ c10::intrusive_ptr<Store> PrefixStore::getUnderlyingStore() {
+      break; // We've reached a non-PrefixStore
+    }
+  }
+  return store;
```
assert store?

fduwjj replied:
You mean null check?

wconstab reviewed (commented):
I think this is OK.  I might slightly prefer the explicit api approach but i would probably stamp this.

fduwjj committed:
Update on "[c10d] Add a recursive method to get the inner most store"


In c10d PG initialization, we wrap TCPStore with multiple layers of PrefixStore which adds layers of prefix.

One example is: 
"default_pg/0//cuda//timeout_dump"
When initialized the default PG, because there is no store passed. We first add the prefix "default_pg" to the TCPStore returned from rendezvous:

https://github.com/pytorch/pytorch/blob/bdeaaad70ccc5a34ef6244b6454d01ca2c582645/torch/distributed/distributed_c10d.py#L1240

We then add pg_name (aka 0) https://github.com/pytorch/pytorch/blob/bdeaaad70ccc5a34ef6244b6454d01ca2c582645/torch/distributed/distributed_c10d.py#L1376 and device (aka cuda) https://github.com/pytorch/pytorch/blob/bdeaaad70ccc5a34ef6244b6454d01ca2c582645/torch/distributed/distributed_c10d.py#L1387

to the prefix. Then when we call store_->set("timeout_dump"). The actual key used for writing into TCPStore is "default_pg/0//cuda//timeout_dump".

For sub-PG, things get even interesting, we put the store wrapped with default pg name to a cache:
https://github.com/pytorch/pytorch/blob/bdeaaad70ccc5a34ef6244b6454d01ca2c582645/torch/distributed/distributed_c10d.py#L1517

And when creating each subPG, it is append its PG name right after the cached store. The example keys are:
'default_pg/0//10//cuda//timeout_dump', 'default_pg/0//12//cuda//timeout_dump', 'default_pg/0//38//cuda//timeout_dump', 'default_pg/0//39//cuda//timeout_dump'. (10, 12, 38 and 39 are all PG names of each subPG created)

The reason why the number in the name is bumped up so high is because for each subPG creation, all ranks have to call the API together and the global variable used for PG name will be bumped up monolithically:

https://github.com/pytorch/pytorch/blob/bdeaaad70ccc5a34ef6244b6454d01ca2c582645/torch/distributed/distributed_c10d.py#L3666

Similar things happen for using hashing for PG names.

This has a potential issue, because each sub-PG has an instance of ProcessGroupNCCL, and if we want to set something global to notify all sub-PGs (and all ranks). This added prefix causes bugs. For example, if on sub-PG 1, we set a value to TCPStore with key ('default_pg/0//1//cuda//timeout_dump'), while we use the default PG instances to check the TCPStore, which are using the key ('default_pg/0//cuda//timeout_dump'), default PG instances will never get the notified signals. So in this PR, we added a new API in PrefixStore which we get the innermost non-PrefixStore for set and check. The next PR will make changes in NCCL watchdog.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
